src/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -172,6 +173,15 @@ public override void Complete(Exception exception = null)
 
             _isReaderCompleted = true;
 
+            // Make an attempt to cancel any call to Stream.ReadAsync
+            _completeCts.Cancel();
```

Comment:
If we're going to allow PipeReader.Complete() to be called during an ongoing call to PipeReader.ReadAsync(), we should ensure that the ongoing ReadAsync calls prior to disposing the inner stream and returning memory pool blocks.

Reply:
This isn’t trying to guard against that, it’s trying to guard against the fact that we have a dangling ReadAsync to the Stream going on, not misuse of the existing API

Reply:
> it’s trying to guard against the fact that we have a dangling ReadAsync to the Stream going on

I know. And that dangling Stream.ReadAsync() call must have been made by a dangling PipeReader.ReadAsync() call. If we don't await the call Stream.ReadAsync() here in Complete before disposing the inner Stream, it's just as much of a misuse of the API as not cancelling the read at all and trying to dispose the inner Stream.

Reply:
No I’m not going to guard against that.

Reply:
This guard happens when the API is being used properly. The other is just buggy code. While it’s easy to misuse I’m not in this PR adding guards for explicit misuse

Reply:
Then we're relying on dispose to do the right thing during an ongoing read anyway. Just remove the call to Cancel().

Reply:
The difference between explicit misuse and implicit misuse is the fact that the StreamPipeReader is yielding reads without canceling the underlying stream. In that case, we want to best effort cancel the pending read on compete since the caller can’t write any code that fixes the situation.

For the case where you fire and forget a call to ReadAsync you just need to fix your code to *not* do that and things will be happy.

Reply:
> the caller can’t write any code that fixes the situation

I'm not sure I understand. The caller should have canceled the read themselves and waited for ReadAsync to complete with a TaskCanceledException. This cancellation *only* handles the case where you fire and forget a call to ReadAsync you just need to fix your code to not do that and things will be happy.

Reply:
Never mind. You're thinking of the CancelPendingRead case.

In that case, I'm even more concerned about properly awaiting the completion of ReadAsyc(), because as you say there's nothing the caller can do.

Reply:
Yea but I’m not willing to block so..... 😬. Either we make this change or we don’t. Also cancellation is a best effort in general so no guarantees this even works without this change (but at least you’re stuck in a ReadAsyc call). The fact that we don’t always dispose the stream is also a problem.

Reply:
We probably should block. It should be rare for there to be any ongoing reads at this point, and in practice, Cancel should cause the read to finish very quickly anyway.

I understand that returning the blocks in the background after Complete exits is not an option. CompleteAsync cannot come soon enough.

Reply:
I rather not introduce sync over async.

Reply:
I’ll open another issue on adding CompleteAsync to both the reader and writer